content: add docker blog (#118)
nextflow-io/website: content/blog/2023/a-nextflow-docker-murder-mystery-the-mysterious-case-of-the-oom-killer.md

Before long, reports of similar mysterious deaths began to trickle in from other jurisdictions. It turned out that the killer had struck [Cromwell](https://cromwell.readthedocs.io/en/stable/) also ([see the police report here](https://github.com/aws/aws-cli/issues/5876)). We breathed a sigh of relief that we could rule out Nextflow as the culprit, but we still had a killer on the loose and a series of container murders to solve!

## Recreating the scene of the crime

As any good detective knows, recreating the scene of the crime is a good place to start. It turned out that our killer had a profile and had been targeting containers processing large datasets since 2020. We came across an excellent [codefresh.io article](https://codefresh.io/blog/docker-memory-usage/) by Saffi Hartal, discussing similar murders and suggesting techniques to lure the killer out of hiding and protect the victims. Unfortunately, the suggested workaround of periodically clearing kernel buffers was impractical in our Nextflow pipeline scenario.

> We borrowed a Python script (`dockertest.py`) from Saffi’s article designed to write huge files and simulate the issues we saw with the Linux buffer and page cache. Using this script, we hoped to replicate the conditions at the time of the murders.

Reviewer: @gwright99 / @GJSissons - is there meant to be a script associated with the filename `dockertest.py` or something? It seems a bit odd having it here. I can't find any mention of the same within Saffi's article either.

Reviewer: I assume it's this python script?

Might be nice to include it in our blog post somewhere, just in case the other site goes down / the link breaks etc. (I would trust a GitHub gist etc more).

```python
import os
import sys
from collections import defaultdict
lines = 0
file_cnt = 0

d=defaultdict(list)
for k, v in ((k.lstrip('-'), v) for k,v in (a.split('=') for a in sys.argv[1:])):
     d[k].append(v)
print (d)
limit = int(d.get('limit')[0]) if  d.get('limit')  else None
same_file =  False if not d.get('bloat') or d.get('bloat')[0] in ['true', 'on'] else True
stats =  True if not d.get('stats') or d.get('stats')[0] in ['true', 'on'] else False
print (f'argv = {sys.argv} limit: {limit} same_file {same_file}  stats {stats}')
# create the file encoded for the yaml
os.system('FNAME="write_lots.py" && cat "$FNAME" | python -m base64 > "$FNAME.b64"')
while(limit is None or lines< limit) : file_cnt += 1 filename= f'file' if same_file else f'bloat{file_cnt}' f = open(f'{filename}', "w") for x in range(1,1000*1000+1): f.write(f'{file_cnt}) {x}: write line {lines} ============================================> \n');
     lines += 1
     if (lines%(1000*1000)==0):
        print(f'Written {file_cnt} {filename}) {x}: write line {lines}')
  f.close()
  print(f'{file_cnt}) close {filename} written {x} total {lines} ');
  if (stats):
     try:
        os.system('''ps auxww | grep -E "python [w]rite_lots|[P]ID " ''')
        os.system('free')
     except(e):
        pass
```

```bash
$ docker run --rm -it -v $PWD/dockertest.py:/dockertest.py --entrypoint /bin/bash --memory="512M" --memory-swap=0 python:3.10.5-slim-bullseye
```

Sure enough, we found that containers began dying with out-of-memory errors. Sometimes we could run a single container, and sometimes we could run two. Containers died even though memory use was well under the cgroups-enforced maximum, as reported by docker stats. As containers ran, we also used the Linux free command to monitor memory usage and the combined memory used by kernel buffers and the page cache.

## Developing a theory of the case

From our testing, we were able to clear both Nextflow and the AWS S3 copy facility since we could replicate the out-of-memory error in our controlled environment independent of both.

We had multiple theories of the case: ***Was it Colonel Mustard with an improper cgroups configuration? Was it Professor Plum and the size of the SWAP partition? Was it Mrs. Peacock running a Linux 5.20 kernel?***

> (For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game)))

Reviewer: ```suggestion
> _(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here(https://cluecluedo.fandom.com/wiki/Clue_(board_game))_
```

- Creating swap files on the Docker host and making swap available to containers using the switch `--memory-swap="1g"` appeared to help, and we learned a lot in the process. Using this workaround we could reliably run 10 containers simultaneously, whereas previously, we could run only one or two. This was a good workaround for static clusters but wasn’t always helpful in cloud batch environments. Creating the swap partition requires root privileges, and in batch environments, where resources may be provisioned automatically, this could be difficult to implement. It also didn’t explain the root cause of why containers were being killed. You can use the commands below to create a swap partition:

```bash
$ sudo dd if=/dev/zero of=/mnt/2GiB.swap bs=2048 count=1048576
$ mkswap /mnt/2GiB.swap
$ swapon /mnt/2GiB.swap
```

## A break in the case!

> On Nov 16th, we finally caught a break in the case. A hot tip from Seqera Lab’s own Jordi Deu-Pons, indicated the culprit may be lurking in the Linux kernel. He suggested hard coding limits for two Linux kernel parameters as follows:

Reviewer: ```suggestion
On Nov 16th, we finally caught a break in the case. A hot tip from Seqera Lab’s own [Jordi Deu-Pons](https://github.com/jordeu), indicated the culprit may be lurking in the Linux kernel. He suggested hard coding limits for two Linux kernel parameters as follows:
```

We ended up working around the problem for Nextflow Tower users by tweaking the kernel parameters in the compute environment deployed by Tower Forge. This solution works reliably with AMIs based on both the 4.14 and 5.10 kernels. We considered adding a swap partition as this was another potential solution to the problem. However, we were concerned that this could have performance implications, particularly for customers running with EBS gp2 magnetic disk storage.

Interestingly, we also tested the [Fusion v2 file system](https://seqera.io/fusion/) with NVMe disk. Using Fusion, we avoided the bug entirely on both kernel versions without needing to adjust kernel partitions or add a swap partition.

## Some helpful investigative tools

If you find evidence of foul play in your cloud or cluster, here are some useful investigative tools you can use:

> - After manually starting a container, use [docker stats](https://docs.docker.com/engine/reference/commandline/stats/) to monitor the CPU and memory used by each container compared to available memory.
>   ```bash
>   $ watch docker stats
>   ```
> - The Linux [free](https://linuxhandbook.com/free-command/) utility is an excellent way to monitor memory usage. You can track total, used, and free memory and monitor the combined memory used by kernel buffers and page cache reported in the *buff/cache* column.
>   ```bash
>   $ free -h
>   ```
> - After a container was killed, we executed the command below on the Docker host to confirm why the containerized Python script was killed.
>   ```bash
>   $ dmesg -T | grep -i ‘killed process’
>   ```
> - We used the Linux [htop](https://man7.org/linux/man-pages/man1/htop.1.html) command to monitor CPU and memory usage to check the results reported by Docker and double-check CPU and memory use.
> - You can use the command [systemd-cgtop](https://www.commandlinux.com/man-page/man1/systemd-cgtop.1.html) to validate group settings and ensure you are not running into arbitrary limits imposed by *cgroups*.
> - Related to the *cgroups* settings described above, you can inspect various memory-related limits directly from the file system. You can also use an alias to make the large numbers associated with *cgroups* parameters easier to read. For example:
>   ```bash
>   $ alias n='numft --to=iec-i'
>   $ cat /sys/fs/cgroup/memory/docker/DOCKER_CONTAINER/memory.limit_in_bytes | n
>   512Mi
>   ```
> - You can clear the kernel buffer and page cache that appears in the buff/cache columns reported by the Linux *free* command using either of these commands:
>   ```bash
>   $ echo 1 > /proc/sys/vm/drop_caches
>   $ sysctl -w vm.drop_caches=1
>   ```

Reviewer: Bulleted code blocks aren't rendering properly - instead showing as inline code. Attempt to fix with additional whitespace and indentation: (needs testing / may need tweaking after merge)
````suggestion
- After manually starting a container, use [docker stats](https://docs.docker.com/engine/reference/commandline/stats/) to monitor the CPU and memory used by each container compared to available memory.

    ```bash
    $ watch docker stats
    ```

- The Linux [free](https://linuxhandbook.com/free-command/) utility is an excellent way to monitor memory usage. You can track total, used, and free memory and monitor the combined memory used by kernel buffers and page cache reported in the *buff/cache* column.

    ```bash
    $ free -h
    ```

- After a container was killed, we executed the command below on the Docker host to confirm why the containerized Python script was killed.

    ```bash
    $ dmesg -T | grep -i ‘killed process’
    ```

- We used the Linux [htop](https://man7.org/linux/man-pages/man1/htop.1.html) command to monitor CPU and memory usage to check the results reported by Docker and double-check CPU and memory use.
- You can use the command [systemd-cgtop](https://www.commandlinux.com/man-page/man1/systemd-cgtop.1.html) to validate group settings and ensure you are not running into arbitrary limits imposed by *cgroups*.
- Related to the *cgroups* settings described above, you can inspect various memory-related limits directly from the file system. You can also use an alias to make the large numbers associated with *cgroups* parameters easier to read. For example:

    ```bash
    $ alias n='numft --to=iec-i'
    $ cat /sys/fs/cgroup/memory/docker/DOCKER_CONTAINER/memory.limit_in_bytes | n
    512Mi
    ```

- You can clear the kernel buffer and page cache that appears in the buff/cache columns reported by the Linux *free* command using either of these commands:

    ```bash
    $ echo 1 > /proc/sys/vm/drop_caches
    $ sysctl -w vm.drop_caches=1
    ```
````

Reviewer: Phew, seems to have worked 😅

title=A Nextflow-Docker Murder Mystery: The mysterious case of the “OOM killer”
date=2023-06-19
type=post
description=Helpful tips for troubleshooting out-of-memory errors in containerized execution environments.
image=img/a-nextflow-docker-murder-mystery-the-mysterious-case-of-the-oom-killer.jpg
tags=nextflow
status=published
author=Graham Wright
icon=graham.jpg
~~~~~~
## Troubleshooting out-of-memory errors with Nextflow and Docker

> *By: Graham Wright, Solutions Architect at Seqera Labs*

Reviewer: I don't think we usually have this line? I think the front matter `author` and `icon` should be sufficient.

```bash
$ docker run --rm -it -v $PWD/dockertest.py:/dockertest.py --entrypoint /bin/bash --memory="512M" --memory-swap=0 python:3.10.5-slim-bullseye
```

Sure enough, we found that containers began dying with out-of-memory errors. Sometimes we could run a single container, and sometimes we could run two. Containers died even though memory use was well under the cgroups-enforced maximum, as reported by docker stats. As containers ran, we also used the Linux free command to monitor memory usage and the combined memory used by kernel buffers and the page cache.

## Developing a theory of the case

From our testing, we were able to clear both Nextflow and the AWS S3 copy facility since we could replicate the out-of-memory error in our controlled environment independent of both.

We had multiple theories of the case: ***Was it Colonel Mustard with an improper cgroups configuration? Was it Professor Plum and the size of the SWAP partition? Was it Mrs. Peacock running a Linux 5.20 kernel?***

> (For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game)))

Reviewer: it's called Cluedo in Europe ;)

```suggestion
(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game)))
```

To make a long story short, we identified several suspects and conducted tests to clear each suspect one by one. Tests included the following:

- We conducted tests with EBS vs. NVMe disk volumes to see if the error was related to page caches when using EBS. The problems persisted with NVMe but appeared to be much less severe.

> - We attempted to configure a swap partition as recommended in this [AWS article](https://repost.aws/knowledge-center/ecs-resolve-outofmemory-errors), which discusses similar out-of-memory errors in Amazon ECS (used by AWS Batch). AWS provides good documentation on managing container [swap space](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-swap.html) using the *--memory-swap* switch. You can learn more about how Docker manages swap space in the [Docker documentation](https://docs.docker.com/config/containers/resource_constraints/).

Reviewer: ```suggestion
- We attempted to configure a swap partition as recommended in this [AWS article](https://repost.aws/knowledge-center/ecs-resolve-outofmemory-errors), which discusses similar out-of-memory errors in Amazon ECS (used by AWS Batch). AWS provides good documentation on managing container [swap space](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-swap.html) using the `--memory-swap` switch. You can learn more about how Docker manages swap space in the [Docker documentation](https://docs.docker.com/config/containers/resource_constraints/).
```

> - Creating swap files on the Docker host and making swap available to containers using the switch *--memory-swap=”1g”* appeared to help, and we learned a lot in the process. Using this workaround we could reliably run 10 containers simultaneously, whereas previously, we could run only one or two. This was a good workaround for static clusters but wasn’t always helpful in cloud batch environments. Creating the swap partition requires root privileges, and in batch environments, where resources may be provisioned automatically, this could be difficult to implement. It also didn’t explain the root cause of why containers were being killed. You can use the commands below to create a swap partition:

Reviewer: ```suggestion
- Creating swap files on the Docker host and making swap available to containers using the switch `--memory-swap="1g"` appeared to help, and we learned a lot in the process. Using this workaround we could reliably run 10 containers simultaneously, whereas previously, we could run only one or two. This was a good workaround for static clusters but wasn’t always helpful in cloud batch environments. Creating the swap partition requires root privileges, and in batch environments, where resources may be provisioned automatically, this could be difficult to implement. It also didn’t explain the root cause of why containers were being killed. You can use the commands below to create a swap partition:
```

> ```

Reviewer: Should be bash and with `$` instead of `#`. Same in the code block below.

By hardcoding these [kernel parameters](https://docs.kernel.org/admin-guide/sysctl/vm.html), we were limiting the number of dirty pages the kernel could hold before writing pages to disk. When these variables were not set, they defaulted to 0, and the default parameters *dirty_ratio* and *dirty_bakground_ratio* took effect instead.

In high-load conditions (such as data-intensive Nextflow pipeline tasks), processes accumulated dirty pages faster than the kernel could flush them to disk, eventually leading to the out-of-memory condition. By hard coding the dirty pages limit, we forced the kernel to flush the dirty pages to disk, thereby avoiding the bug. This also explained why the problem was less pronounced using NVMe storage, where flushing to disk occurred more quickly, thus mitigating the problem.

Further testing determined that the bug appeared reliably on the newer [AMI Linux 2 AMI using the 5.10 kernel](https://aws.amazon.com/about-aws/whats-new/2021/11/amazon-linux-2-ami-kernel-5-10/). The bug did not seem to appear when using the older Amazon Linux 2 AMI running the 4.14 kernel version.

We now had two solid strategies to resolve the problem and thwart our killer:

> - Create a swap partition and run containers with the *--memory-swap* flag set.

Reviewer: ```suggestion
- Create a swap partition and run containers with the `--memory-swap` flag set.
```

> - Set *dirty_bytes* and *dirty_background_bytes* kernel variables on the Docker host before launching the jobs.

Reviewer: ```suggestion
- Set `dirty_bytes` and `dirty_background_bytes` kernel variables on the Docker host before launching the jobs.
```

Before long, reports of similar mysterious deaths began to trickle in from other jurisdictions. It turned out that the killer had struck [Cromwell](https://cromwell.readthedocs.io/en/stable/) also ([see the police report here](https://github.com/aws/aws-cli/issues/5876)). We breathed a sigh of relief that we could rule out Nextflow as the culprit, but we still had a killer on the loose and a series of container murders to solve!

## Recreating the scene of the crime

As any good detective knows, recreating the scene of the crime is a good place to start. It turned out that our killer had a profile and had been targeting containers processing large datasets since 2020. We came across an excellent [codefresh.io article](https://codefresh.io/blog/docker-memory-usage/) by Saffi Hartal, discussing similar murders and suggesting techniques to lure the killer out of hiding and protect the victims. Unfortunately, the suggested workaround of periodically clearing kernel buffers was impractical in our Nextflow pipeline scenario.

> We borrowed a Python script (`dockertest.py`) from Saffi’s article designed to write huge files and simulate the issues we saw with the Linux buffer and page cache. Using this script, we hoped to replicate the conditions at the time of the murders.

Reviewer: ```suggestion
We borrowed the Python script from Saffi’s article designed to write huge files and simulate the issues we saw with the Linux buffer and page cache. Using this script, we hoped to replicate the conditions at the time of the murders.
```

```bash
$ docker run --rm -it -v $PWD/dockertest.py:/dockertest.py --entrypoint /bin/bash --memory="512M" --memory-swap=0 python:3.10.5-slim-bullseye
```

Sure enough, we found that containers began dying with out-of-memory errors. Sometimes we could run a single container, and sometimes we could run two. Containers died even though memory use was well under the cgroups-enforced maximum, as reported by docker stats. As containers ran, we also used the Linux free command to monitor memory usage and the combined memory used by kernel buffers and the page cache.

## Developing a theory of the case

From our testing, we were able to clear both Nextflow and the AWS S3 copy facility since we could replicate the out-of-memory error in our controlled environment independent of both.

We had multiple theories of the case: ***Was it Colonel Mustard with an improper cgroups configuration? Was it Professor Plum and the size of the SWAP partition? Was it Mrs. Peacock running a Linux 5.20 kernel?***

> _(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here(https://cluecluedo.fandom.com/wiki/Clue_(board_game))_

Reviewer: Try again 🤦🏻 

```suggestion
_(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game)))_
```

> _For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game))_

Reviewer: It must be something with the `)` in the URL. Trying wikipedia instead..

```suggestion
_For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://en.wikipedia.org/wiki/Cluedo)_
```

> > _(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here(https://cluecluedo.fandom.com/wiki/Clue_(board_game))_

Reviewer: oof sorry, blockquote does not look good in this theme.

```suggestion
_(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here(https://cluecluedo.fandom.com/wiki/Clue_(board_game))_
```

> _(For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game)))_

Reviewer: Don't understand why the website doesn't like this, it's fine on GitHub. Try without the `()`.
```suggestion
_For the millennials and Gen Zs in the crowd, you can find a primer on the CLUE/Cluedo references [here](https://cluecluedo.fandom.com/wiki/Clue_(board_game))_
```
